src/test/regress/sql/union_pushdown.sql

```diff
@@ -230,5 +230,297 @@ SELECT DISTINCT user_id FROM
+			JOIN
+		((SELECT * FROM users_table_part JOIN events_table_part USING (user_id)) UNION ALL (SELECT * FROM users_table_part JOIN events_table_part USING (user_id))) as l2
+			USING(user_id))
+	) as left_subquery
+	UNION ALL
+	SELECT * FROM
+	(
+		(((SELECT * FROM users_table_part JOIN events_table_part USING (user_id)) UNION ALL (SELECT * FROM users_table_part JOIN events_table_part USING (user_id))) as l1
+			JOIN
+		((SELECT * FROM users_table_part JOIN events_table_part USING (user_id)) UNION ALL (SELECT * FROM users_table_part JOIN events_table_part USING (user_id))) as l2
+			USING(user_id))
+	) as right_subquery
+) as top_level_righy USING (user_id)
+ORDER BY user_id DESC
+LIMIT 1;
+
+-- a subquery in WHERE clause with
+-- a tree with   [Q1.1 JOIN Q1.2 UNION ALL Q2.1 JOIN Q2.2] JOIN [Q3.1 JOIN Q3.2 UNION ALL Q4.1 JOIN Q4.2]
+-- can be pushed down with FROM tree consisting of JOINs/UNION ALLs
+SELECT * FROM
+users_table_part u1
+	JOIN
+events_table_part e1 USING (user_id)
+	JOIN
+users_table_part u2 USING (user_id)
+	JOIN
+(SELECT * FROM users_table_part UNION ALL SELECT * FROM events_table_part) as foo USING (user_id)
+WHERE user_id IN
+	(SELECT user_id FROM (
+	SELECT * FROM
+	(
+		(((SELECT * FROM users_table_part JOIN events_table_part USING (user_id)) UNION ALL (SELECT * FROM users_table_part JOIN events_table_part USING (user_id))) as l1
+			JOIN
+		((SELECT * FROM users_table_part JOIN events_table_part USING (user_id)) UNION ALL (SELECT * FROM users_table_part JOIN events_table_part USING (user_id))) as l2
+			USING(user_id))
+	) as left_subquery
+	UNION ALL
+	SELECT * FROM
+	(
+		(((SELECT * FROM users_table_part JOIN events_table_part USING (user_id)) UNION ALL (SELECT * FROM users_table_part JOIN events_table_part USING (user_id))) as l1
+			JOIN
+		((SELECT * FROM users_table_part JOIN events_table_part USING (user_id)) UNION ALL (SELECT * FROM users_table_part JOIN events_table_part USING (user_id))) as l2
+			USING(user_id))
+	) as right_subquery
+) as top_level_left
+	JOIN
+(
+	SELECT * FROM
+	(
+		(((SELECT * FROM users_table_part JOIN events_table_part USING (user_id)) UNION ALL (SELECT * FROM users_table_part JOIN events_table_part USING (user_id))) as l1
+			JOIN
+		((SELECT * FROM users_table_part JOIN events_table_part USING (user_id)) UNION ALL (SELECT * FROM users_table_part JOIN events_table_part USING (user_id))) as l2
+			USING(user_id))
+	) as left_subquery
+	UNION ALL
+	SELECT * FROM
+	(
+		(((SELECT * FROM users_table_part JOIN events_table_part USING (user_id)) UNION ALL (SELECT * FROM users_table_part JOIN events_table_part USING (user_id))) as l1
+			JOIN
+		((SELECT * FROM users_table_part JOIN events_table_part USING (user_id)) UNION ALL (SELECT * FROM users_table_part JOIN events_table_part USING (user_id))) as l2
+			USING(user_id))
+	) as right_subquery
+) as top_level_righy USING (user_id)
+ORDER BY user_id DESC
+) ORDER BY 1 LIMIT 1;
```

Comment:
If the `events_table_part` is a reference table I get this error:

`+ERROR:  no distribution column found for relation 17051, because it is a reference table`

This comes from the join order but is this expected? Also we should probably give the table name in this error


```
CREATE TABLE events_table_part(user_id bigint, value_1 int, value_2 int) ;
-- CREATE TABLE events_table_part_0 PARTITION OF events_table_part FOR VALUES FROM (0) TO (1);
-- CREATE TABLE events_table_part_1 PARTITION OF events_table_part FOR VALUES FROM (1) TO (2);
-- CREATE TABLE events_table_part_2 PARTITION OF events_table_part FOR VALUES FROM (2) TO (3);
-- CREATE TABLE events_table_part_3 PARTITION OF events_table_part FOR VALUES FROM (3) TO (4);
-- CREATE TABLE events_table_part_4 PARTITION OF events_table_part FOR VALUES FROM (4) TO (5);
-- CREATE TABLE events_table_part_5 PARTITION OF events_table_part FOR VALUES FROM (5) TO (6);
-- CREATE TABLE events_table_part_6 PARTITION OF events_table_part FOR VALUES FROM (6) TO (7);
-- CREATE TABLE events_table_part_7 PARTITION OF events_table_part FOR VALUES FROM (7) TO (8);
-- CREATE TABLE events_table_part_8 PARTITION OF events_table_part FOR VALUES FROM (8) TO (9);
SELECT create_reference_table('events_table_part');
```

Reply:
And if this is a citus local table we get:

` relation events_table_part is not distributed`

Reply:
for reference we fixed this, and added regression tests